crates/cext/src/transpiler/transpile_layout.rs

```diff
@@ -0,0 +1,337 @@
+        let _ = Box::into_raw(box_slice);
+        QkLayoutArray { size, array }
+    } else {
+        QkLayoutArray {
+            size: 0,
+            array: std::ptr::null_mut(),
+        }
+    }
+}
+
+/// @ingroup QkTranspileLayout
+/// Return the routing permutation of a ``QkTranspileLayout``
+///
+/// @param layout A pointer to the ``QkTranspileLayout``.
+///
+/// @returns A QkLayoutArray object which is a size and pointer to an array of the routing
+/// permutation. This will need to be passed to ``qk_layout_array_free()`` to free the array allocation.
+///
+/// # Safety
+/// Behavior is undefined if ``layout`` is not a valid, non-null pointer to a
+/// ``QkTranspileLayout``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_transpile_layout_output_permutation(
+    layout: *const TranspileLayout,
+) -> QkLayoutArray {
+    let layout = unsafe { const_ptr_as_ref(layout) };
+    let permutation = layout.explicit_output_permutation().to_vec();
+    let mut box_slice = permutation.into_boxed_slice();
+    let array = box_slice.as_mut_ptr() as *mut u32;
+    let size = box_slice.len();
+    let _ = Box::into_raw(box_slice);
+    QkLayoutArray { size, array }
+}
+
+/// @ingroup QkTranspileLayout
+/// Return the final layout of a ``QkTranspileLayout``
+///
+/// @param layout A pointer to the ``QkTranspileLayout``.
+/// @param filter_ancillas If set to true the output array will not include any indicies for any
+/// ancillas added by the transpiler.
+///
+/// @returns A QkLayoutArray object which is a size and pointer to an array of the final layout. This
+/// will need to be passed to ``qk_layout_array_free()`` to free the array allocation.
+///
+/// # Safety
+/// Behavior is undefined if ``layout`` is not a valid, non-null pointer to a
+/// ``QkTranspileLayout``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_transpile_layout_final_layout(
+    layout: *const TranspileLayout,
+    filter_ancillas: bool,
+) -> QkLayoutArray {
+    let layout = unsafe { const_ptr_as_ref(layout) };
+    let final_layout = layout.final_index_layout(filter_ancillas);
+    let mut box_slice = final_layout.into_boxed_slice();
+    let array = box_slice.as_mut_ptr() as *mut u32;
+    let size = box_slice.len();
+    let _ = Box::into_raw(box_slice);
+    QkLayoutArray { size, array }
+}
+
+/// @ingroup QkTranspileLayout
+/// Free a QkLayoutArray
+///
+/// This function will free the memory allocation for an array pointed to by a ``QkLayoutArray``.
+///
+/// @param layout_array The QkLayoutArray object to free the array allocation of.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``layout_array`` is not an QkLayoutArray object created by either
+/// ``qk_transpile_layout_final_layout()``, ``qk_transpile_layout_output_permutation()``, or
+/// ``qk_transpile_layout_initial_layout()``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_layout_array_free(layout_array: QkLayoutArray) {
```

Comment:
In our current setup `..._free` always frees a pointer, while `_clear` clears out a struct (like for `QkCircuitInstruction`). Here we're doing the latter, right?